pkg/virt-handler/node-labeller/kvm-caps-info-plugin.go

```diff
@@ -0,0 +1,263 @@
+    struct kvm_msr_list *msrs = (struct kvm_msr_list*)data;
+    if (n >= msrs->nmsrs) {
+        return 1;
+    }
+    *index = msrs->indices[n];
+    return 0;
+}
+*/
+import "C"
+
+import (
+	"os"
+	"syscall"
+	"unsafe"
+
+	"kubevirt.io/client-go/log"
+)
+
+const (
+	HV_X64_MSR_CRASH_CTL               = 0x40000105
+	HV_X64_MSR_RESET                   = 0x40000003
+	HV_X64_MSR_VP_INDEX                = 0x40000002
+	HV_X64_MSR_VP_RUNTIME              = 0x40000010
+	HV_X64_MSR_SCONTROL                = 0x40000080
+	HV_X64_MSR_STIMER0_CONFIG          = 0x400000B0
+	HV_X64_MSR_TSC_FREQUENCY           = 0x40000022
+	HV_X64_MSR_REENLIGHTENMENT_CONTROL = 0x40000106
+	HV_X64_MSR_VP_ASSIST_PAGE          = 0x40000073
+)
+
+type tmsr struct {
+	nmsrs   C.__u32
+	indices []C.__u32
+}
+
+type CapScan struct {
+	supportedMSRS map[uint32]bool
+	fd            uintptr
+	logger        *log.FilteredLogger
+}
+
+type capability struct {
+	Name      string
+	Extension uintptr
+	MSR       int
+}
+
+func NewCapScanner() *CapScan {
+	capScan := new(CapScan)
+	capScan.logger = log.DefaultLogger()
+	return capScan
+}
+
+var CapsDesc = []capability{
+	{
+		Extension: uintptr(C.CapHyperv),
+		Name:      "base",
+	},
+	{
+		Extension: uintptr(C.CapHypervTime),
+		Name:      "time",
+	},
+	{
+		Extension: uintptr(C.CapHypervVpIndex),
+		MSR:       HV_X64_MSR_VP_INDEX,
+		Name:      "vpindex",
+	},
+	{
+		Extension: uintptr(C.CapHypervTlbflush),
+		Name:      "tlbflush",
+	},
+	{
+		Extension: uintptr(C.CapHypervSendIPI),
+		Name:      "ipi",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic2),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic2",
+	},
+	{
+		MSR:  HV_X64_MSR_TSC_FREQUENCY,
+		Name: "frequencies",
+	},
+	{
+		MSR:  HV_X64_MSR_RESET,
+		Name: "reset",
+	},
+	{
+		MSR:  HV_X64_MSR_VP_RUNTIME,
+		Name: "runtime",
+	},
+	{
+		MSR:  HV_X64_MSR_STIMER0_CONFIG,
+		Name: "synictimer",
+	},
+	{
+		MSR:  HV_X64_MSR_REENLIGHTENMENT_CONTROL,
+		Name: "reenlightenment",
+	},
+}
+
+func availableMsrs(fd uintptr) ([]uint32, error) {
+
+	// Find our list of MSR indices.
+	// A page should be more than enough here,
+	// eventually if it's not we'll end up with
+	// a failed system call for some reason other
+	// than E2BIG (which just says n is wrong).
+	msrIndices := make([]byte, 4096, 4096)
+	msrs := make([]uint32, 0, 0)
+	for {
+
+		_, err := kvmIOCtl(
+			fd,
+			uintptr(C.IoctlGetMsrIndexList),
+			uintptr(unsafe.Pointer(&msrIndices[0])))
+		if err == syscall.E2BIG {
+			// The nmsrs field will now have been
+			// adjusted, and we can run it again.
+			continue
+		} else if err != 0 {
+			return nil, err
+		}
+
+		// We're good!
+		break
+	}
+
+	// Extract each msr individually.
+	for i := 0; ; i++ {
+		// Is there a valid index?
+		var index C.__u32
+		e := C.msr_list_index(
+			unsafe.Pointer(&msrIndices[0]),
+			C.int(i),
+			&index)
```

Comment:
Instead of calling this function, it should be possible to reinterpret the buffer `msrIndices` as a pointer to `C.kvm_msr_list` structure:
```
msrStruct := (*C.kvm_msr_list)(unsafe.Pointer(&msrIndices[0]))
```

They you could access the structure directly in go code. I have not tired to compile it, so maybe it will not work.

Reply:
Ignore my previous comment, it would not work as expected because the C structure has dynamic size.
Instead can you use 2 small C functions to access it?
```
int msr_list_length(void* data) {
	struct kvm_msr_list *msrs = (struct kvm_msr_list*)data;
	return msrs->nmsrs
}

__u32 msr_list_get(void* data, int index) {
	struct kvm_msr_list *msrs = (struct kvm_msr_list*)data;
	return msrs->indices[index];
}
```

It would be more readable than using the one that is there now.